<a href="https://colab.research.google.com/github/chavamoon/MachineLearningExamples/blob/main/Neural_networks_and_pytorch_intro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import torch

## **Pytorch intro**

#### Empty torch torch.empty 

Returns a tensor filled with uninitialized data. The shape of the tensor is defined by the variable argument size.
https://pytorch.org/docs/stable/generated/torch.empty.html

In [6]:
# Scalar
x = torch.empty(1)
print('\n----------------------------Scalar----------------------------:\n', x)
print('Size:', x.size())

#Vector
x = torch.empty(1,4)
print('\n----------------------------Vector----------------------------:\n', x)
print('Size:', x.size())

#Matrix
x = torch.empty(3,3)
print('\n----------------------------Matrix----------------------------:\n', x)
print('Size:', x.size())

#3d Tensor
x = torch.empty(3,5,4)
print('\n----------------------------3D Tensor----------------------------:\n', x)
print('Size:', x.size())


----------------------------Scalar----------------------------:
 tensor([6.8126e+19])
Size: torch.Size([1])

----------------------------Vector----------------------------:
 tensor([[6.8127e+19, 3.0848e-41, 3.3631e-44, 0.0000e+00]])
Size: torch.Size([1, 4])

----------------------------Matrix----------------------------:
 tensor([[6.8125e+19, 3.0848e-41, 3.3631e-44],
        [0.0000e+00,        nan, 6.4460e-44],
        [4.4721e+21, 1.5956e+25, 4.7399e+16]])
Size: torch.Size([3, 3])

----------------------------3D Tensor----------------------------:
 tensor([[[2.0090e+18, 3.0848e-41, 2.8727e-43, 0.0000e+00],
         [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
         [0.0000e+00, 0.0000e+00, 1.9422e+31, 2.7491e+20],
         [6.1949e-04, 1.9421e+31, 2.7491e+20, 2.3078e-12],
         [1.8788e+31, 7.9303e+34, 6.1949e-04, 1.8590e+34]],

        [[7.7767e+31, 7.1536e+22, 1.8180e+31, 1.4580e-19],
         [1.1495e+24, 3.0956e-18, 4.7851e+22, 4.1575e+21],
         [8.3186e+20, 1.707

#### Random, ones and zeros tensors

In [7]:
print(torch.rand(3,4))
print(torch.ones(3,4))
print(torch.zeros(3,4))

tensor([[0.1725, 0.1760, 0.5230, 0.2637],
        [0.2441, 0.5798, 0.0397, 0.1798],
        [0.9468, 0.5599, 0.0830, 0.5049]])
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])


#### Tensor types

In [8]:
#double
x= torch.rand(3,2, dtype=torch.double)
print('\nDouble: \n', x)

#float
x= torch.rand(3,2, dtype=torch.float)
print('\nFloat: \n', x)


Double: 
 tensor([[0.9720, 0.9280],
        [0.0990, 0.7677],
        [0.2182, 0.4357]], dtype=torch.float64)

Float: 
 tensor([[0.1291, 0.8198],
        [0.2982, 0.9236],
        [0.0854, 0.0635]])


#### From List to tensor

In [11]:
print('\nTensor from list \n')
x = torch.tensor([3.0,4.0,5.0])
print(x)
print(x.size())

print('\nTensor from list of lists (vector)\n')
x = torch.tensor([[3.0,4.0,5.0]])
print(x)
print(x.size())

print('\nTensor from list of lists (matrix)\n')
x = torch.tensor([[3.0,4.0,5.0],[1.0, 2.0, 3.0]])
print(x)
print(x.size())


Tensor from list 

tensor([3., 4., 5.])
torch.Size([3])

Tensor from list of lists (vector)

tensor([[3., 4., 5.]])
torch.Size([1, 3])

Tensor from list of lists (matrix)

tensor([[3., 4., 5.],
        [1., 2., 3.]])
torch.Size([2, 3])


#### Tensor operations

##### Tensor * Scalar

In [12]:
x = torch.ones(2,2)
x *3

tensor([[3., 3.],
        [3., 3.]])

##### Tensors sum

Two ways, with + symbol or with torch.add method

In [14]:
x = 2*torch.ones(2,2)
y = torch.ones(2,2)

print(x)
print(y)

print('\nSum with + \n', x+y)
print('\nSum with pytorch add\n', torch.add(x,y))


tensor([[2., 2.],
        [2., 2.]])
tensor([[1., 1.],
        [1., 1.]])

Sum with + 
 tensor([[3., 3.],
        [3., 3.]])

Sum with pytorch add
 tensor([[3., 3.],
        [3., 3.]])


Inplace sum with add_
All torch operations contains an inplace wersion with a method with _ suffix 

In [15]:
y = torch.ones(2,2)
print(y)
z = y
z.add_(x)
print('\n',y)

tensor([[1., 1.],
        [1., 1.]])

 tensor([[3., 3.],
        [3., 3.]])


##### Copy tensors

Tensors are passed by reference by new variables, if you want to create a copy of a torch the easiest way is with clone().detach()

In [16]:
y = torch.ones(2,2)
print(y)
z = y.clone().detach()
z.add_(x)
print('\n',y)

tensor([[1., 1.],
        [1., 1.]])

 tensor([[1., 1.],
        [1., 1.]])


##### Tensor multiplication

mul() for element-wise multiplication  
div for element-wise div
If you add _ suffix the operation will be done inplace

In [18]:
x = 2* torch.ones(2,2)
y = 3* torch.ones(2,2)
print(x)
print(y)
print('\n mul\n', torch.mul(x,y))
print('\n div\n', torch.div(x,y))

tensor([[2., 2.],
        [2., 2.]])
tensor([[3., 3.],
        [3., 3.]])

 mul
 tensor([[6., 6.],
        [6., 6.]])

 div
 tensor([[0.6667, 0.6667],
        [0.6667, 0.6667]])


#### Tensor slices

In [20]:
x = torch.rand(2,3)
print(x)
print('\nFirst column:\n', x[:,0])
print('\nSecond column:\n',x[:,1])
print('\nFirst row:\n', x[0,:])
print('\nFirst element:\n', x[0,0])

tensor([[0.0926, 0.1827, 0.4434],
        [0.7679, 0.1382, 0.1795]])

First column:
 tensor([0.0926, 0.7679])

Second column:
 tensor([0.1827, 0.1382])

First row:
 tensor([0.0926, 0.1827, 0.4434])

First element:
 tensor(0.0926)


#### Tensor reshape

**view method allow us to modify tensors shape**

view(-1) reduces one dimension

In [21]:
x = torch.rand(4,4)
print(x)

print('\n Unidimensional tensor: \n')
print(x.view(-1))

tensor([[0.5726, 0.7866, 0.1130, 0.8924],
        [0.9729, 0.6911, 0.1535, 0.4172],
        [0.2860, 0.7404, 0.6883, 0.3342],
        [0.2379, 0.2465, 0.3115, 0.6316]])

 Unidimensional tensor: 

tensor([0.5726, 0.7866, 0.1130, 0.8924, 0.9729, 0.6911, 0.1535, 0.4172, 0.2860,
        0.7404, 0.6883, 0.3342, 0.2379, 0.2465, 0.3115, 0.6316])


If we use with -1 argument with the desired size from the result dimension, torch will calculate it automatically.
For example:
 * 8 columns (-1,8)
 * 8 rows (8,-1)

In [22]:
print(x)
print('\n 8 columns tensor : \n')
print(x.view(-1, 8))

print('\n 8 rows tensor : \n')
print(x.view(8, -1))

tensor([[0.5726, 0.7866, 0.1130, 0.8924],
        [0.9729, 0.6911, 0.1535, 0.4172],
        [0.2860, 0.7404, 0.6883, 0.3342],
        [0.2379, 0.2465, 0.3115, 0.6316]])

 8 columns tensor : 

tensor([[0.5726, 0.7866, 0.1130, 0.8924, 0.9729, 0.6911, 0.1535, 0.4172],
        [0.2860, 0.7404, 0.6883, 0.3342, 0.2379, 0.2465, 0.3115, 0.6316]])

 8 rows tensor : 

tensor([[0.5726, 0.7866],
        [0.1130, 0.8924],
        [0.9729, 0.6911],
        [0.1535, 0.4172],
        [0.2860, 0.7404],
        [0.6883, 0.3342],
        [0.2379, 0.2465],
        [0.3115, 0.6316]])


**None argument allow us to modify dimensions too**

[None,:] Adds a extra dimesion where Nonw keyword is found. It helps us when we multiply matrixs

In [23]:
y = torch.tensor([1.0,1.0,1.0])
print(y)
print('Dim : ', y.shape)

tensor([1., 1., 1.])
Dim :  torch.Size([3])


In [27]:
y = torch.tensor([1.0,1.0,1.0])[:,None]
print(y)
print('Dim : ', y.shape)
y2 = y[:,None]
print(y2)
print('Dim : ', y2.shape)

tensor([[1.],
        [1.],
        [1.]])
Dim :  torch.Size([3, 1])
tensor([[[1.]],

        [[1.]],

        [[1.]]])
Dim :  torch.Size([3, 1, 1])


In [25]:
y = torch.tensor([1.0,1.0,1.0])[None,:]
print(y)
print('Dim : ', y.shape)

tensor([[1., 1., 1.]])
Dim :  torch.Size([1, 3])


#### Numpy & Pytorch

##### Pytorch -> Numpy

with numpy() method. It works as a pointer. Updates un torch object modifies numpy. Only when the tensor is not in GPU

In [28]:
import numpy as np

a = torch.ones(3)
print(a)
b = a.numpy()
print(b)

a.add_(1)
print(a)
print(b)


tensor([1., 1., 1.])
[1. 1. 1.]
tensor([2., 2., 2.])
[2. 2. 2.]


##### Numpy -> Torch

In [29]:
a = np.ones((3,2))
print(a)
b = torch.from_numpy(a)
print(b)


[[1. 1.]
 [1. 1.]
 [1. 1.]]
tensor([[1., 1.],
        [1., 1.],
        [1., 1.]], dtype=torch.float64)


#### Matrix multiplication

In [30]:
#Element wise with mul or *
#Matrix product with matmul

#1.0 for converting matrix to float type

x = 1.0*torch.tensor([[1,1,1],[2,2,2],[3,3,3]])
print(x)
y = torch.eye(3,3)
print(y)

print('\n Element wise multiplication : \n', x*y)

print('\n Matrix product : \n', torch.matmul(x,y))

tensor([[1., 1., 1.],
        [2., 2., 2.],
        [3., 3., 3.]])
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]])

 Element wise multiplication : 
 tensor([[1., 0., 0.],
        [0., 2., 0.],
        [0., 0., 3.]])

 Matrix product : 
 tensor([[1., 1., 1.],
        [2., 2., 2.],
        [3., 3., 3.]])


In [39]:
x = torch.tensor([[1.0,1.0,1.0],[2.0,2.0,2.0],[3.0,3.0,3.0]])
print('x:\n',x)
print('Dim:', x.shape)

y = torch.tensor([1.0,1.0,1.0])[:,None]
print('y:\n',y)
print('Dim:', y.shape)

print('mul x, y:\n', x.mul(y))
print('Matmul x, y:\n', x.matmul(y))

x:
 tensor([[1., 1., 1.],
        [2., 2., 2.],
        [3., 3., 3.]])
Dim: torch.Size([3, 3])
y:
 tensor([[1.],
        [1.],
        [1.]])
Dim: torch.Size([3, 1])
mul x, y:
 tensor([[1., 1., 1.],
        [2., 2., 2.],
        [3., 3., 3.]])
Matmul x, y:
 tensor([[3.],
        [6.],
        [9.]])


#### Transpose

In [31]:
x = torch.tensor([[1.0,1.0,1.0],[2.0,2.0,2.0],[3.0,3.0,3.0]])
print(x)
print('\n Matriz Transpuesta \n')
x.transpose(0,1)

tensor([[1., 1., 1.],
        [2., 2., 2.],
        [3., 3., 3.]])

 Matriz Transpuesta 



tensor([[1., 2., 3.],
        [1., 2., 3.],
        [1., 2., 3.]])

## **Pytorch neural network intro**

nn component help us to define neural networks

In [40]:
from torch import nn

#### Neural Network definition as a class

We define a neuron with an input layer, a hidden layer and a output layer with feedforwading architecture

In [45]:
class NeuralNetwork(nn.Module):

  def __init__(self, in_n=784, hidden_n=156, out_n=1):
    super(NeuralNetwork,self).__init__()

    # w0:= Input weights -> hidden layer
    # nn.Linear(input neurons, hidden_neurons, bias=True)
    self.hidden = nn.Linear(in_n, hidden_n)

    # W1:= Hidden layer weights -> output layer  
    # nn.Linear(hidden neurons, output neurons, bias=True)
    self.output = nn.Linear(hidden_n, out_n)
        
    # Activation functions 
    # Sigma 1 [hidden layer]
    self.sigmoid = nn.Sigmoid()
    # Sigma 2 [output layer]
    self.softmax = nn.Softmax(dim=1)
  
  def forward(self, x):
    # x0 * w0
    x=self.hidden(x)
    # sigma1 evaluation, get x1
    x = self.sigmoid(x)

    # x1 * w1
    x = self.output(x)
    # Sigma2 evaluation, get y
    x = self.softmax(x)
    return x

Doing the feedforward.
Weights and bias are initialized automatically

In [46]:
# Seed
torch.manual_seed(345)

# 28x28 = 784 (Imagine an 28x28 image)
x0 = torch.rand((1, 784))
model_nn = NeuralNetwork(in_n=784, hidden_n=256, out_n=1)
resultado = model_nn(x0)
print(resultado)

tensor([[1.]], grad_fn=<SoftmaxBackward0>)


For evaluating our network we try with a smaller network and we change softwax for an identity function

In [48]:
torch.manual_seed(345)


n_in = 3
n_hid = 5
n_out = 1

x0 = torch.rand((1, n_in))

model_nn = NeuralNetwork(in_n=n_in, hidden_n=n_hid, out_n=n_out)
model_nn.softmax = nn.Identity()
resultado = model_nn(x0)
print(resultado)

tensor([[0.3952]], grad_fn=<AddmmBackward0>)


Evaluating the network

with .weight we can check the layer weights

In [49]:
# Hidden layer weight
print(model_nn.hidden.weight)

Parameter containing:
tensor([[ 0.3351, -0.4106, -0.3619],
        [-0.4247, -0.3092,  0.0131],
        [ 0.4064,  0.1796,  0.4533],
        [ 0.4600,  0.2685,  0.0652],
        [ 0.0783, -0.5293,  0.5048]], requires_grad=True)


In [50]:
# Hidden layer input
W0 = model_nn.hidden.weight
b0 = model_nn.hidden.bias[None,:]
# No olvides utilizar None para hacer las dimensiones compatibles
print('Dim W0:   ', W0.shape)
print('Dim b0:   ', b0.shape)

print()

# Output layer input
W1 = model_nn.output.weight

b1 = model_nn.output.bias[None,:] 
print('Dim W1:   ', W1.shape)
print('Dim b1:   ', b1.shape)

# x0 dimension
print('\nDim de x0 : ',x0.shape)

Dim W0:    torch.Size([5, 3])
Dim b0:    torch.Size([1, 5])

Dim W1:    torch.Size([1, 5])
Dim b1:    torch.Size([1, 1])

Dim de x0 :  torch.Size([1, 3])


Dimensions are correct but weight matrix are transposed

Manual evaluation

In [51]:
# Input from hidden layer
input1 = torch.matmul(x0, W0.transpose(0,1))
input1 = input1 + b0
print(input1.shape)

# Exit from hidden layer
sigma_1 = nn.Sigmoid()
hidden_out = sigma_1(input1)
hidden_out.shape

torch.Size([1, 5])


torch.Size([1, 5])

In [52]:
# Input from output layer
input2 = torch.matmul(hidden_out, W1.transpose(0,1))
input2 = input2 + b1
print(input2.shape)
# network output
sigma_2 = nn.Identity()
y_out = sigma_2(input2)
y_out[0]

torch.Size([1, 1])


tensor([0.3952], grad_fn=<SelectBackward0>)

Results from network and manual evaluation are thea same .3952

Setting weights manually

In [53]:
torch.manual_seed(345)


n_in = 3
n_hid = 5
n_out = 1

model_nn = NeuralNetwork(in_n=n_in, hidden_n=n_hid, out_n=n_out)

# Definamos los pesos manualmente
x0 = torch.rand((1, n_in))
W0_ = torch.rand((n_hid, n_in))
b0_ = torch.rand((n_hid))
W1_ = torch.rand((n_out, n_hid))
b1_ = torch.rand((n_out))

#How to set weights manually.
model_nn.hidden.weight.data = W0_
model_nn.hidden.bias.data = b0_
model_nn.output.weight.data = W1_
model_nn.output.bias.data = b1_

resultado = model_nn(x0)
print(resultado)

tensor([[1.]], grad_fn=<SoftmaxBackward0>)


#### Neural network with sequential

In [54]:

n_x = 784
n_hidden = [256]
n_y = 1

# Neural network with sequential
model_s = nn.Sequential(
    # X0 x W0 
    nn.Linear(n_x, n_hidden[0]), 
    # X1 = sigma_1(W0 x X0)
    nn.Sigmoid(),
    # X1 x W1
    nn.Linear(n_hidden[0], n_y),
    # Y = X2 = sigma_2(W1 x X1)
    nn.Softmax(dim=1))

print(model_s)

Sequential(
  (0): Linear(in_features=784, out_features=256, bias=True)
  (1): Sigmoid()
  (2): Linear(in_features=256, out_features=1, bias=True)
  (3): Softmax(dim=1)
)
